source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -850,19 +850,16 @@ are parsed from the ismaster response in the obvious way.
 Server Description Equality
 ```````````````````````````
 
-For the purposes of updating topology description and publishing SDAM events,
-two server descriptions having the same address
-MUST be considered equal if and only if the values of
-`ServerDescription`_ fields marked (=) are respectively equal.
+For the purpose of determining whether to publish SDAM events, two server
+descriptions having the same address MUST be considered equal if and only if
+the values of `ServerDescription`_ fields marked (=) are respectively equal.
+Two topology descriptions MUST be considered equal if and only if they have
+the same set of server addresses, and the server descriptions for each address
+are respectively equal as just defined.
```

Comment:
[SDAM Monitoring: Events](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-monitoring.rst#events) states the following rules for dispatching events:

> ServerDescriptionChangedEvent: When the old server description is not equal to the new server description
> TopologyDescriptionChangedEvent: When the old topology description is not equal to the new topology description.

Given that, perhaps this text is sufficient to clarify that events would not be dispatched.

Beyond that, do we need to remind implementations that SDAM flow must be run even if server descriptions are considered "equal" under the above rules? I believe that's main point of SPEC-1517, and simply removing "purposes of updating topology description and..." from the first sentence may be too subtle a change to communicate that.

[Parsing an ismaster response](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#parsing-an-ismaster-response) already states:

> This replacement MUST happen even if the new server description compares equal to the previous one

I think that's clear for the server description case; however, [Updating the TopologyDescription](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#updating-the-topologydescription) does not have similar language requiring some action even if server/topology descriptions are equal. I think the new sentence talking about topology description equality would be relevant, since that previously wasn't defined/addressed.

@mbroadst: do you concur?

Reply:
Yes, and if I'm reading this correctly then I think that's what @jyemin is alluding to above. Perhaps we should define `TopologyDescription` equality, and add clarifying language to `Updating the TopologyDescription`

Reply:
I removed language pertaining to topology equality from this PR.

In the [sdam spec](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#topologydescription), topology description is not defined to contain any servers. This language matches Ruby implementation in that the topology object only contains topology-related attributes as they are described in the referenced section, and does not contain server descriptions.

[sdam events spec](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-monitoring.rst#events) then prescribes that topology change event is published when:

> When the old topology description is not equal to the new topology description.

Based on this, my reading is when the set of servers in a topology changes, but the topology type remains the same, a topology description change event must NOT be published. However, I believe this reading contradicts the spec tests such as [this one](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/tests/monitoring/replica_set_with_primary.yml#L27) that appears to require a topology change event to be published when no topology attribute changed.

If there is agreement that the current definition of topology description and existing monitoring spec tests are contradictory, I propose opening a new spec ticket to resolve the contradiction. SPEC-1517 refers to always processing ismaster responses, which has much smaller scope than resolving discrepancy between topology description definition and existing spec tests (as well as, presumably, driver implementations).

Reply:
If I follow correctly, https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/tests/monitoring/replica_set_with_primary.yml#L27 is a problem because the test expects a `topology_description_changed_event` event when `topologyType` remains "Unknown" and only the list of servers has changed. I agree that seems odd and also agree that it seems independent of this PR and would warrant its own spec ticket.

Reply:
I created https://jira.mongodb.org/browse/SPEC-1567.